kernel/src/vm/vmar/vm_mapping.rs

```diff
@@ -218,50 +219,53 @@ impl VmMapping {
         self.inner.lock().map_size += extra_size;
     }
 
-    pub fn handle_page_fault(
-        &self,
-        page_fault_addr: Vaddr,
-        not_present: bool,
-        write: bool,
-    ) -> Result<()> {
-        let required_perm = if write { VmPerms::WRITE } else { VmPerms::READ };
-        self.check_perms(&required_perm)?;
+    pub fn handle_page_fault(&self, page_fault_info: &PageFaultInfo) -> Result<()> {
+        self.check_perms(&page_fault_info.required_perms)?;
+
+        let address = page_fault_info.address;
+
+        let page_aligned_addr = address.align_down(PAGE_SIZE);
+        let is_write = page_fault_info.required_perms.contains(VmPerms::WRITE);
 
-        if !write && self.vmo.is_some() && self.handle_page_faults_around {
-            self.handle_page_faults_around(page_fault_addr)?;
+        if !is_write && self.vmo.is_some() && self.handle_page_faults_around {
+            self.handle_page_faults_around(address)?;
             return Ok(());
         }
 
-        let page_aligned_addr = page_fault_addr.align_down(PAGE_SIZE);
-
-        if write && !not_present {
-            // Perform COW at page table.
-            let root_vmar = self.parent.upgrade().unwrap();
-            let mut cursor = root_vmar
-                .vm_space()
-                .cursor_mut(&(page_aligned_addr..page_aligned_addr + PAGE_SIZE))?;
-            let VmItem::Mapped {
+        let root_vmar = self.parent.upgrade().unwrap();
+        let mut cursor = root_vmar
+            .vm_space()
+            .cursor_mut(&(page_aligned_addr..page_aligned_addr + PAGE_SIZE))?;
+        match cursor.query()? {
+            VmItem::Mapped {
                 va: _,
                 frame,
                 mut prop,
-            } = cursor.query().unwrap()
-            else {
-                return Err(Error::new(Errno::EFAULT));
-            };
-
-            if self.is_shared {
-                cursor.protect(PAGE_SIZE, |p| p.flags |= PageFlags::W);
-            } else {
-                let new_frame = duplicate_frame(&frame)?;
-                prop.flags |= PageFlags::W;
-                cursor.map(new_frame, prop);
+            } if is_write => {
+                if self.is_shared {
+                    cursor.protect(PAGE_SIZE, |p| p.flags |= PageFlags::W);
+                } else {
+                    let new_frame = duplicate_frame(&frame)?;
```

Comment:
In this case we need to copy on write. Recently an optimization #1158 is introduced to safe the copy if the current process is the only owner (which happens if the child exit very early). You need to restore it.